juju/model.py

```diff
@@ -152,6 +195,28 @@ def _load(self):
         return self._backend.relation_get(self.relation_id, self.unit.name)
 
 
+class MutableRelationUnitData(RelationUnitData, MutableLazyMapping):
```

Comment:
The complexity of the class graph doesn't seem to be paying off. Here is a suggestion: before committing to this structure, I suggest doing a spike where these new types go away and we keep just one single `RelationUnitData` with a `self._mutable` boolean defining whether to allow writes. Just put the logic directly in `__setitem__` and `__getitem__` themselves. My gut feeling is that we'll need a trivial number of lines on each of these, and all of that additional complexity will disappear.